### DISCUSSION TITLE
How to allocate python operator resources

### DISCUSSION THREAD

**Discussion by Lee2532** [apache/airflow#39538] [Q&A]
I'm using airflow in my Kubernetes environment

Some tasks allocate resources through kpo, but some tasks run through the basic Python operator without doing anything else. However, sometimes those jobs cause OOM, and I want to allocate resources in advance to prevent this, is there a way?

**Comment by obragaa** [comment]
Hii Lee2532,

In your Airflow setup on Kubernetes, if you're experiencing out-of-memory (OOM) issues with tasks running through the basic Python operator, you can manage resource allocation by defining resource requests and limits directly in your task definitions.

When you define a task in Airflow using the KubernetesPodOperator, you can specify resource requests and limits as part of the pod's configuration. However, for tasks that utilize the basic PythonOperator, you need to ensure these are managed at the Kubernetes level, typically through Kubernetes Executor configs if you're using this executor.

Here's an example of how you might set this up in your DAG:

```python
from airflow.decorators import dag, task
from airflow.utils.dates import days_ago
from airflow.kubernetes.pod import Resources

default_args = {
    'owner': 'airflow',
}

@dag(default_args=default_args, schedule_interval=None, start_date=days_ago(2))
def resource_managed_dag():
    @task(resources={'request_memory': '500Mi', 'request_cpu': '500m', 'limit_memory': '1000Mi', 'limit_cpu': '1000m'})
    def process_data():
        # Python code to process data here
        pass

    process_data()

dag = resource_managed_dag()
```

In this example, the `@task` decorator is used to define memory and CPU requests and limits. Adjust these values based on your specific workload requirements.

It's also recommended to monitor your tasks' resource usage to adjust these settings optimally over time. Tools like Kubernetes metrics-server or Prometheus can be helpful for this.

**Reply by Lee2532** [reply]
It was requests_memory.
I wrote requests only

thanks :)

**Reply by smktpd** [reply]
Where did you take this import (`from airflow.kubernetes.pod import Resources`) from? Because of this answer now all AI parrot this bullshit.

**Reply by smktpd** [reply]
Answering myself: this was actual for a prehistoric version [2.2.0 of cncf-kubernetes provider via fqcn `airflow.providers.cncf.kubernetes.backcompat.pod.Resources`](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/2.2.0/_api/airflow/providers/cncf/kubernetes/backcompat/pod/index.html), which since then has long been deprecated.

[Airflow docs](https://airflow.apache.org/docs/apache-airflow/2.11.0/tutorial/taskflow.html#dependency-separation-using-kubernetes-pod-operator) suggest using `@task.kubernetes` decorator ([code](https://github.com/apache/airflow/blob/providers-cncf-kubernetes/10.12.3/providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/decorators/kubernetes.py)) that [since ver 5.0.0 of cncf-kubernetes provider](https://github.com/apache/airflow/blob/providers-cncf-kubernetes/10.12.3/providers/cncf/kubernetes/docs/changelog.rst#500) [has param](https://github.com/apache/airflow/blob/providers-cncf-kubernetes/10.12.3/task-sdk/src/airflow/sdk/definitions/decorators/__init__.pyi#L526) `container_resources` that accepts [objects](https://github.com/apache/airflow/blob/providers-cncf-kubernetes/10.12.3/airflow-core/src/airflow/example_dags/example_kubernetes_executor.py#L195-L198) of type `k8s.V1ResourceRequirements`
